tools/rosmaster/src/rosmaster/authorization.py

```diff
@@ -0,0 +1,482 @@
+
+            if "parameters" in doc.keys():
+                for k, v in doc["parameters"].items():
+                    if k in self.reserved_parameters:
+                        raise ImportError('Parameter %s is a reserved topic!' % k )
+                    if "setters" in v.keys():
+                        self.setters[k] = self.resolve( v["setters"] )
+                    else:
+                        raise ImportError('Parameter %s does not have setters!' % k )
+                    if "getters" in v.keys():
+                        self.getters[k] = self.resolve( v["getters"] ) | self.setters[k]
+                    else:
+                        self.logger.warn( 'Parameter %s does not have getters!' % k )
+            else:
+                self.logger.warn( "Authorization file %s does not contain parameters" % config_file )
+                pass
+        else:
+            raise ImportError( "Authorization file is empty" )
+
+        """ set pub/sub for reserved topics """
+
+        self.publishers["/rosout"] = set( ip for ips in chain( self.publishers.values(), self.subscribers.values() ) for ip in ips ) 
+        self.subscribers["/rosout"] = self.master 
+
+        self.publishers["/rosout_agg"] = self.master
+        self.subscribers["/rosout_agg"] = self.master
+
+        """ set ip_address for reserved nodes """
+
+        self.nodes["/rosout"] = self.master
+
+        """ compute peer publishers for each subscriber IP addresses """
+        topics = set( self.publishers.keys() + self.subscribers.keys() )
+        self.peer_publishers = { s: set() for subs in self.subscribers.values() for s in subs }
+        for t in topics:
+            for s in self.subscribers[t]:
+                self.peer_publishers[s].update( set( self.publishers[t] ) )
+
+        """ compute all authorized ip addresses """
+
+        self.ip_addresses = set( ip for ip in chain.from_iterable( self.nodes.values() ) )
+        self.ip_addresses.update( set( ip for ip in chain.from_iterable( self.publishers.values() ) ) )
+        self.ip_addresses.update( set( ip for ip in chain.from_iterable( self.subscribers.values() ) ) )
+        self.ip_addresses.update( set( ip for ip in chain.from_iterable( self.providers.values() ) ) )
+        self.ip_addresses.update( set( ip for ip in chain.from_iterable( self.requesters.values() ) ) )
+        self.ip_addresses.update( set( ip for ip in chain.from_iterable( self.setters.values() ) ) )
+        self.ip_addresses.update( set( ip for ip in chain.from_iterable( self.getters.values() ) ) )
+
+        """ set setters/getters for reserved parameters """
+        for p in self.reserved_parameters:
+            self.setters[p] = self.master
+            self.getters[p] = self.ip_addresses
+        
+
+    def save( self, filename ):
+        """ Write full authorization configuration to YAML file
+        """
+        data = OrderedDict()
```

Comment:
if we're targeting Noetic that means python3, dicts are ordered by default so this is unnecessary. Also, why not just one call to update with one big dictionary?

Also still noticing lots of extra spaces around params and curly braces, not sure if this is a problem with the style of the existing code.